asf_public_discourse_home_decarbonisation/analysis/mse/initial_text_analysis_category_data.py

```diff
@@ -0,0 +1,442 @@
+        "ngrams_col": "3_grams_text",
+        "var_used": "posts",
+        "min_frequency": min_frequency_trigrams,
+    },
+}
+
+
+def process_text_and_titles(mse_data: pd.DataFrame) -> pd.DataFrame:
+    """
+    Apply basic processing to text data and identifying stopwords.
+
+    Args:
+        mse_data (pd.DataFrame): Money Saving Expert data
+
+    Returns:
+        pd.DataFrame: MSE data with processed text
+    """
+    mse_data["processed_text"] = mse_data["text"].apply(lambda x: process_text(x))
+    mse_data["processed_title"] = mse_data["title"].apply(lambda x: process_text(x))
+    return mse_data
+
+
+def create_a_mapping_between_tokens_and_lemmas(mse_data: pd.DataFrame) -> dict:
+    """
+    Create a mapping between tokens and lemmas for each token in either text or titles
+    Args:
+        mse_data (pd.DataFrame): Money Saving Expert data
+
+    Returns:
+        dict: a dictionary mapping tokens to lemmatised tokens
+    """
+    # Create a list of tokens in titles (removing repetitions from replies, which have the same title)
+    tokens_title = mse_data[mse_data["is_original_post"] == 1][["tokens_title"]]
+    tokens_title = tokens_title.explode("tokens_title").rename(
+        columns={"tokens_title": "tokens"}
+    )
+
+    # Create a list of tokens in text
+    tokens_text = mse_data[["tokens_text"]]
+    tokens_text = tokens_text.explode("tokens_text").rename(
+        columns={"tokens_text": "tokens"}
+    )
+
+    # Concatenate both lists
+    all_tokens = pd.concat([tokens_text, tokens_title])
+
+    # Unique tokens
+    all_tokens = all_tokens.drop_duplicates("tokens")
+
+    # Removing digits as they cannot be lemmatized
+    # (needs to explicitly to have `== False` because is_digit can also be missing)
+    all_tokens["is_digit"] = all_tokens["tokens"].str.isdigit()
+    all_tokens = all_tokens[all_tokens["is_digit"] == False]
+
+    lemmas_dictionary = lemmatize_sentence(list(all_tokens["tokens"]))
+
+    return lemmas_dictionary
+
+
+def tokenize_and_lemmatise_data(mse_data: pd.DataFrame) -> pd.DataFrame:
+    """
+    Tokenizes and lemmatises text data.
+
+    Args:
+        mse_data (pd.DataFrame): Money Saving Expert data
+
+    Returns:
+        pd.DataFrame: Money Saving Expert data with tokenized and lemmatised added columns
+    """
+
+    # Tokenizing and lemmatising text and titles variables
+    mse_data["tokens_title"] = mse_data["processed_title"].apply(word_tokenize)
+    mse_data["tokens_text"] = mse_data["processed_text"].apply(word_tokenize)
+
+    # Creating a list with all tokens (either in title or text of posts/replies)
+    # and creating a dictonary that maps tokens to lemmatised tokens:
+    lemmas_dictionary = create_a_mapping_between_tokens_and_lemmas(mse_data)
+
+    # Lemmatising tokens - using np.vectorize seems to be the fastest way of lemmatizing
+    # if lemma is not found, the original word is kept
+    replace_words_vec = np.vectorize(
```

Comment:
I have a conceptual question, when doing the `lemmas_dictionary.get(word,word)` operation, I am guessing some of the words could be the same but given the context that they are used in then they may have different lemmas? How does this distinguish between those cases?

Reply:
Unless I am understanding this wrong I think this relates to [Liz's comment above](https://github.com/nestauk/asf_public_discourse_home_decarbonisation/pull/11/files/ca7af5b612e0702bb2d4a59be2b01b03c55eb89f#r1417819917).

I was originally applying lemmatisation to each post/reply individually
Given the problems with the size of the dataset I thought of how I could speed things up so I started doing this... which is obviously problematic;
But even the first approach is not 100% correct, so I'll apply lemmatisation sentence by sentence. Does it sound okay @helloaidank?

Reply:
Never replied properly to this, but sounds good to me!